server/lib/src/utils.rs

```diff
@@ -85,12 +88,69 @@ impl Distribution<char> for DistinctAlpha {
     }
 }
 
+pub(crate) struct GraphemeClusterIter<'a> {
+    value: &'a str,
+    char_bounds: Vec<usize>,
+    window: usize,
+    range: Range<usize>,
+}
+
+impl<'a> GraphemeClusterIter<'a> {
+    pub fn new(value: &'a str, window: usize) -> Self {
+        let char_bounds = if value.len() < window {
+            Vec::with_capacity(0)
+        } else {
+            let mut char_bounds = Vec::with_capacity(value.len());
+            for idx in 0..value.len() {
+                if value.is_char_boundary(idx) {
+                    char_bounds.push(idx);
+                }
+            }
+            char_bounds.push(value.len());
```

Comment:
Manual reimplementation of filter :thinking:

```suggestion
            let mut char_bounds: Vec<usize> = (0..(value.len())).filter(|idx| value.is_char_boundary(*idx)).collect();
```